### PR TITLE
Adds support for pushing helm charts

### PR DESCRIPTION
**Pull request by JosephSalisbury**
Towards https://github.com/giantswarm/giantswarm/issues/1253

- [x] add helm commands to workflow
- [x] test out helm commands
- [x] template `Chart.yaml` in build
- [x] test api chart build
- [x] chart support docs

### CODE REVIEW

File: workflow/workflow_test.go [hunk 1]
```diff
@@ -410,7 +430,7 @@ func TestGetDeployWorkflow(t *testing.T) {
 					index,
 					testIndex,
 					expectedCommandName,
-					workflow[index].Name,
+					workflow[testIndex].Name,
```
Comment (JosephSalisbury) [inline workflow/workflow_test.go]: This is me fixing a faulty test

**Comment by JosephSalisbury** [comment]
See https://github.com/giantswarm/api/pull/384/files for an example of a chart

**Comment by JosephSalisbury** [comment]
See https://github.com/giantswarm/giantswarm/wiki/Architect-Helm-Support for architect helm support docs

File: template/template.go [hunk 2]
```diff
@@ -33,6 +35,64 @@ type TemplateConfiguration struct {
 	configuration.Installation
 }
 
+// TemplateHelmChart takea a filesystem, a path to a directory containing a
```
Comment (xh3b4sd) [inline template/template.go]: `takea` -> `takes`

File: template/template.go [hunk 3]
```diff
@@ -33,6 +35,64 @@ type TemplateConfiguration struct {
...
+// helm chart, and a BuildInfo struct.
+// It templates the chart's Chart.yaml and templates/deployment.yaml
+// with this information.
+// It is an error if there are multiple charts in the helm directory.
+func TemplateHelmChart(fs afero.Fs, helmPath string, buildInfo BuildInfo) error {
+	fileInfos, err := afero.ReadDir(fs, helmPath)
+	if err != nil {
+		return err
```
Comment (xh3b4sd) [inline template/template.go]: Please mask the errors.

File: template/template.go [hunk 4]
```diff
@@ -33,6 +35,64 @@ type TemplateConfiguration struct {
...
+	}
+
+	if len(fileInfos) == 0 {
+		return nil
+	}
+
+	if len(fileInfos) > 1 {
+		return fmt.Errorf("Multiple charts found in helm path")
```
Comment (xh3b4sd) [inline template/template.go]: Please use typed errors.

File: template/template.go [hunk 5]
```diff
@@ -33,6 +35,64 @@ type TemplateConfiguration struct {
...
+	}
+
+	chartDirectory := fileInfos[0].Name()
+
+	chartsYamlPath := filepath.Join(helmPath, chartDirectory, "Chart.yaml")
+	deploymentPath := filepath.Join(helmPath, chartDirectory, "templates", "deployment.yaml")
```
Comment (xh3b4sd) [inline template/template.go]: We may should use constants for the magic strings and have proper godoc for them.

File: template/template_test.go [hunk 6]
```diff
@@ -16,6 +16,139 @@ import (
...
+	tests := []struct {
+		helmPath  string
+		buildInfo BuildInfo
+		setUp     func(afero.Fs, string) error
+		check     func(afero.Fs, string) error
+	}{
+		// Test that an empty helm directory does nothing.
+		{
+			helmPath:  "/helm",
+			buildInfo: BuildInfo{},
```
Comment (xh3b4sd) [inline template/template_test.go]: You want to test that the build directory is empty. Here also the build information is empty. IMO this test should have build information provided to have the test case more strictly aligned to its matter. Having both information missing does not verify which one causes the actual denial of work.

File: workflow/helm.go [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+	"fmt"
+	"os/user"
+	"path/filepath"
+
+	"github.com/giantswarm/architect/commands"
+	"github.com/spf13/afero"
+)
+
+const (
+	HelmImage = "quay.io/giantswarm/docker-helm:006b0db51ec484be8b1bd49990804784a9737ece"
```
Comment (xh3b4sd) [inline workflow/helm.go]: I feel like this should be a flag.

File: template/template.go (hunk 3 above)
Comment (JosephSalisbury) [inline template/template.go]: I have some boyscouting to do. See https://github.com/giantswarm/architect/issues/34.

File: template/template.go (hunk 4 above)
Comment (JosephSalisbury) [inline template/template.go]: See https://github.com/giantswarm/architect/issues/34. If this is not my next PR to `architect`, you have permission to beat me <3

Comment (JosephSalisbury) [inline template/template.go]: See https://github.com/giantswarm/architect/issues/34. If this is not my next PR to `architect`, you have permission to beat me <3

File: workflow/helm.go (hunk 7 above)
Comment (JosephSalisbury) [inline workflow/helm.go]: Uncertain. We don't support kubectl via flag. I don't really see the use case.

Comment (JosephSalisbury) [inline workflow/helm.go]: I've actually been considering reducing the number of flags. 99% of the time we don't use flags. I think the only ones we actively use are organisation, and resources directory path.

File: template/template.go [hunk 8]
```diff
@@ -33,6 +46,64 @@ type TemplateConfiguration struct {
...
+			return err
+		}
+
+		if exists {
+			contents, err := afero.ReadFile(fs, path)
+			if err != nil {
+				return err
+			}
+
+			t := template.Must(template.New(path).Funcs(filters).Parse(string(contents)))
```
Comment (kopiczko) [inline template/template.go]: I don't think `.Must` is a good idea when reading template from external source.

File: template/template.go [hunk 9]
```diff
@@ -33,6 +46,64 @@ type TemplateConfiguration struct {
...
+				return err
+			}
+
+			var buf bytes.Buffer
+			if err := t.Execute(&buf, buildInfo); err != nil {
+				return err
+			}
+
+			templatedContents := buf.String()
+			if err := afero.WriteFile(fs, path, []byte(templatedContents), permission); err != nil {
```
Comment (kopiczko) [inline template/template.go]: Why not buf.Bytes()? You need a copy there?

File: template/template.go (hunk 5 above)
Comment (JosephSalisbury) [inline template/template.go]: Done.

File: template/template_test.go (hunk 6 above)
Comment (JosephSalisbury) [inline template/template_test.go]: Good shout. Done.

File: template/template.go [hunk 10]
```diff
@@ -33,6 +46,64 @@ type TemplateConfiguration struct {
...
+		return err
+	}
+
+	if len(fileInfos) == 0 {
+		return nil
+	}
+
+	if len(fileInfos) > 1 {
+		return fmt.Errorf("Multiple charts found in helm path")
+	}
```
Comment (kopiczko) [inline template/template.go]: Probably good to make assertion if `fileInfos[0]` is a directory.

File: template/template_test.go [hunk 11]
```diff
@@ -16,6 +16,141 @@ import (
...
+
+				return nil
+			},
+			check: func(fs afero.Fs, helmPath string) error {
+				fileInfos, err := afero.ReadDir(fs, helmPath)
+				if err != nil {
+					return err
+				}
+				if len(fileInfos) != 0 {
+					return fmt.Errorf("multiple files found in helm directory")
```
Comment (kopiczko) [inline template/template_test.go]: I think it should say "helm directory not empty" or sth. What if len(fileInfos) == 1?

File: template/template_test.go [hunk 12]
```diff
@@ -16,6 +16,141 @@ import (
...
+
+				ingressPath := filepath.Join(helmPath, "test-chart", HelmTemplateDirectoryName, "ingress.yaml")
+				if err := afero.WriteFile(
+					fs,
+					ingressPath,
+					[]byte("host: {{ .Values.Installation.etc }}"),
+					permission,
+				); err != nil {
+					return err
+				}
```
Comment (kopiczko) [inline template/template_test.go]: Please create slice for these files and invoke filepath.Join & afero.WriteFile once in a loop <3

File: template/template_test.go [hunk 13]
```diff
@@ -16,6 +16,141 @@ import (
...
+
+				ingressPath := filepath.Join(helmPath, "test-chart", HelmTemplateDirectoryName, "ingress.yaml")
+				ingressBytes, err := afero.ReadFile(fs, ingressPath)
+				if err != nil {
+					return err
+				}
+				if string(ingressBytes) != "host: {{ .Values.Installation.etc }}" {
+					return fmt.Errorf("values not found in ingress, found: '%v'", string(ingressBytes))
+				}
```
Comment (kopiczko) [inline template/template_test.go]: Again: please create slice, invoke filepath.Join & afero.ReadFile once.

**Pull request merged**